### ISSUE TITLE
Converged range check for material properties

### ISSUE THREAD

**Issue by tophmatthews** [idaholab/moose#14996]
## Reason
There are times when it would be nice to ensure that inputs going into material models are within applicable ranges (e.g. temperature is within an applicable range for a certain thermal conductivity model). In addition, a check on the final converged material property value would help prevent unphysical results (e.g. thermal conductivity is not negative). The primary problem is that during each nonlinear iterations, values may be slightly out the range due to the JFNK solve (I think). Really, it is the final value that should be checked. Then, an InputParameter to decide what to do should be available to do nothing, show a warning, throw and error, or throw an exception.

## Design
I imagine an interface should do the job okay.

```
checkValueBelow(value, upper_limit, intermediate_evaluation)
checkValueAbove(value, lower_limit, intermediate_evaluation)
checkValueBetween(value, lower_limit, upper_limit, intermediate_evaluation)
```

`Value`: number to check
`lower_limit`: lower allowed limit
`upper_limit`: upper allowed
`intermediate_evaluation`: bool to run checks at every evaluation if true, only at timestep end of false.

The methods (e.g. `checkValueBetween`) could do something like
```
if (!intermediate_evaluation && _current_execute_flag != EXEC_TIMESTEP_END)
  return;
if (_behavior != LimitOptions::nothing)
{
  std::ostringstream msg;
  msg << "In "
      << _name << ": number (" << value
      << ") is outside the range (" << lower_limit << " " << upper_limit << ");
  if (_behavior == LimitOptions::warning)
    mooseWarning(msg.str());
  else if (_behavior == LimitOptions::exception)
    mooseException(msg.str());
  else
    mooseError(msg.str());
}
```

The methods (e.g. `checkValueBetween`) could be placed directly in computeQpProperties(). I think they are calculated at the timestep end, but I am not positive.

The two main problems are:
- if computeQpProperties is not run after the variables are converged at the end of a timestep, how can I check final converged values.
- How to throw an exception for a converged value?

## Impact
Better value checking.

**Comment by tophmatthews** [comment]
What do you guys think about this @dschwen @bwspenc @lindsayad ? I think I've talked with you each separately about this at some point...

**Comment by lindsayad** [comment]
seems reasonable to me

**Comment by tophmatthews** [comment]
what about 
> How to throw an exception for a converged value?

I know you thought it wasn't clear how to do that in the past @lindsayad. explicitly, throw an exception and cut the time step at the end of the timestep?

**Comment by lindsayad** [comment]
I don’t think you’d want to throw an exception. If you’re doing a post solve check of the material properties it seems like that could go into a postSolve method in the transient executioner or the time stepper where it would be easy to tell the system to cut the time step.

**Comment by tophmatthews** [comment]
Closed by #23005 (I think)
